### ISSUE TITLE
[Paper Note] On Layer Normalization in the Transformer Architecture, Ruibin Xiong+, arXiv'20, 2020.02

### ISSUE THREAD

**Issue by AkihikoWatanabe** [AkihikoWatanabe/paper_notes#2142]
# URL
- https://arxiv.org/abs/2002.04745
# Authors
  - Ruibin Xiong
  - Yunchang Yang
  - Di He
  - Kai Zheng
  - Shuxin Zheng
  - Chen Xing
  - Huishuai Zhang
  - Yanyan Lan
  - Liwei Wang
  - Tie-Yan Liu
# Abstract
  - The Transformer is widely used in natural language processing tasks. To train a Transformer however, one usually needs a carefully designed learning rate warm-up stage, which is shown to be crucial to the final performance but will slow down the optimization and bring more hyper-parameter tunings. In this paper, we first study theoretically why the learning rate warm-up stage is essential and show that the location of layer normalization matters. Specifically, we prove with mean field theory that at initialization, for the original-designed Post-LN Transformer, which places the layer normalization between the residual blocks, the expected gradients of the parameters near the output layer are large. Therefore, using a large learning rate on those gradients makes the training unstable. The warm-up stage is practically helpful for avoiding this problem. On the other hand, our theory also shows that if the layer normalization is put inside the residual blocks (recently proposed as Pre-LN Transformer), the gradients are well-behaved at initialization. This motivates us to remove the warm-up stage for the training of Pre-LN Transformers. We show in our experiments that Pre-LN Transformers without the warm-up stage can reach comparable results with baselines while requiring significantly less training time and hyper-parameter tuning on a wide range of applications.
# Translation (by gpt-5.6-luna)
- Transformerは、自然言語処理のタスクで広く利用されている。しかし、Transformerを学習するには通常、慎重に設計された学習率ウォームアップ段階が必要である。この段階は最終的な性能に不可欠であることが示されている一方で、最適化を遅らせ、ハイパーパラメータの調整を増やす。本論文ではまず、学習率ウォームアップ段階がなぜ重要なのかを理論的に検討し、レイヤー正規化の位置が重要であることを示す。具体的には、平均場理論を用いて、初期化時において、残差ブロック間にレイヤー正規化を配置する従来設計のPost-LN Transformerでは、出力層に近いパラメータの期待勾配が大きくなることを証明する。そのため、これらの勾配に対して大きな学習率を用いると、学習が不安定になる。ウォームアップ段階は、実際上、この問題を回避するうえで有用である。

一方で、我々の理論は、レイヤー正規化を残差ブロックの内部に配置した場合（近年提案されたPre-LN Transformer）、初期化時の勾配が適切な状態になることも示している。これを受けて、Pre-LN Transformerの学習ではウォームアップ段階を取り除く。実験により、ウォームアップ段階を用いないPre-LN Transformerは、幅広い応用において、ベースラインと同等の結果を達成しながら、学習時間とハイパーパラメータ調整の負担を大幅に削減できることを示す。
# Summary (by gpt-5.6-luna)
- Transformerの学習で必要とされる学習率ウォームアップについて、Layer Normalizationの位置に着目して理論的に分析。  
従来のPost-LN Transformerでは出力層付近の勾配が大きくなり学習が不安定になるため、ウォームアップが有効である。  
一方、Pre-LN Transformerでは初期勾配が適切に保たれるため、ウォームアップなしでも同等の性能を達成し、学習時間とハイパーパラメータ調整を削減できる。

**Comment by AkihikoWatanabe** [comment]
OpenReview:https://openreview.net/forum?id=B1x8anVFPr

**Comment by AkihikoWatanabe** [comment]
Encoder-DecoderのTransformerにおいて、Post-LNの場合は、Warmupを無くすと最終的な性能が悪化し、またWarmUpステップの値によって（500 vs. 4000で実験)も最終的な性能が変化する。これには学習時にハイパーパラメータをしっかり探索しなければならず、WarmUPを大きくすると学習効率が落ちるというデメリットがある。
![image](https://github.com/user-attachments/assets/e7a26ecd-7905-4e6c-bb9a-29b8289addb0)

Post-LNの場合は、Pre-LNと比較して勾配が大きく、Warmupのスケジュールをしっかり設計しないと大きな勾配に対して大きな学習率が適用され学習が不安定になる。これは学習率を非常に小さくし、固定値を使うことで解決できるが、収束が非常に遅くなるというデメリットがある。
![image](https://github.com/user-attachments/assets/afb09f44-c7c9-44ab-9066-3ee788ebd8ee)

一方、Pre-LNはWarmup無しでも、高い性能が達成でき、上記のようなチューニングの手間や学習効率の観点から利点がある、みたいな話の模様。
![image](https://github.com/user-attachments/assets/d675a58b-e876-4e41-a76f-306c2e1ce23f)
